### PR TITLE
fix: add custom (de)serialization methods for special float value

### PR DESCRIPTION
**Pull request by dqhl76**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary
fixes: #16213
- Implemented custom serialization and deserialization for F32 and F64 types in NumberScalar to handle infinite/Nan float values.

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [x] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16258)

### CODE REVIEW

**Comment by andylokandy** [comment]
Thank you!

File: tests/sqllogictests/suites/base/issues/issue_16213.test
Comment (sundy-li) [reply tests/sqllogictests/suites/base/issues/issue_16213.test:8]:
```
🐳 :) insert into a select 'inf' from numbers(100);
100 rows written in 0.200 sec. Processed 100 rows, 8B (500.73 rows/s, 40B/s)

🐳 :) insert into a select 'inf' from numbers(100);
100 rows written in 0.122 sec. Processed 100 rows, 8B (820.21 rows/s, 65B/s)

🐳 :) insert into a select 'nan' from numbers(100);
100 rows written in 0.107 sec. Processed 100 rows, 8B (933.3 rows/s, 74B/s)

🐳 :) select  max(a) from a;
error: APIError: ResponseError with 4000: invalid data (For portability reasons we do not allow to serialize NaNs.)

```

**Comment by andylokandy** [comment]
@sundy-li Is supporting infinite for all NumberScalar a goo idea? As there is no standard for inf serialization, I'd concern about the forward compatibility because NumberScalar could be persist in meta settings. 

On the other aspect, this is an user interface change. We used to / planned to support `'inf'::float64` as the recommended way to construct a inf float. If we accept this PR, it'll implying that we will support auto cast from  string to numeric types, which is denied currently.

**Comment by dqhl76** [comment]
> I'd concern about the forward compatibility because NumberScalar could be persist in meta or settings.

I add a condition `serializer.is_human_readable()` to handle the `inf` case. I find seems we use `bincode` and `msgpack` in meta related(they are not human readable but json is). It could be not affect?

**Comment by sundy-li** [comment]
> I find seems we use bincode and msgpack in meta related(they are not human readable but json is). It could be not affect? Not sure about that 0.0

Yes.

> As there is no standard for inf serialization, I'd concern about the forward compatibility because NumberScalar could be persist in meta or settings.

@andylokandy  If `inf`or `nan` is already serialized into meta or settings via `serde_json`, we can't deserialize it back, it could throw errors, so this pr can solve the problem.


Now we can also cast string into float64 (but it's not auto-cast)
```
🐳 :) select '+inf'::float64, '-inf'::float64, 'nan'::float64, 'Infinity'::float64, '-Infinity'::float64, 'Nan'::float64;
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ '+inf'::float64 │ '-inf'::float64 │ 'nan'::float64 │ 'infinity'::float64 │ '-infinity'::float64 │ 'nan'::float64 │
│     Float64     │     Float64     │     Float64    │       Float64       │        Float64       │     Float64    │
├─────────────────┼─────────────────┼────────────────┼─────────────────────┼──────────────────────┼────────────────┤
│             inf │            -inf │            NaN │                 inf │                 -inf │            NaN │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
1 row read in 0.048 sec. Processed 1 row, 1B (21.05 rows/s, 21B/s)
```

**Review by sundy-li — APPROVED** [review]

File: tests/sqllogictests/suites/base/issues/issue_16213.test
Comment (dqhl76) [reply tests/sqllogictests/suites/base/issues/issue_16213.test:8]: When I try to add the tests, I found that `error: APIError: ResponseError with 4000: invalid data (For portability reasons we do not allow to serialize NaNs.)` still occurs.
This error is issuesed by `borsh`, we use it at `src/query/functions/src/aggregates/adaptors/aggregate_null_unary_adaptor.rs:201:9`. It is not controlled by our overried (de)serde methods.

File: tests/sqllogictests/suites/base/issues/issue_16213.test
Comment (andylokandy) [inline tests/sqllogictests/suites/base/issues/issue_16213.test:10]: May be better change it it `query`

Comment (dqhl76) [reply tests/sqllogictests/suites/base/issues/issue_16213.test:10]: Got that

File: tests/sqllogictests/suites/base/issues/issue_16213.test
Comment (andylokandy) [inline tests/sqllogictests/suites/base/issues/issue_16213.test:12]: Why this is not `inf`?

Comment (dqhl76) [reply tests/sqllogictests/suites/base/issues/issue_16213.test]: Sorry for misunderstanding, fixed

**Comment by dqhl76** [comment]
Sorry for some time leave.

This PR was put on hold because overriding the (de)serialization methods for F32 and F64 still doesn't address the issue. There are parts of the code, such as ValueType::Scalar, that use borsh-serde but are unaffected by this PR's override methods.
img_v3_02dt_20858391-51ef-4763-9a71-7b2b6fbbf30g

To work around this problem, I patched borsh-rs to remove the NaN restriction: [borsh-rs NaN bypass patch](https://github.com/dqhl76/borsh-rs/tree/bypass-nan?rev=6609a44). It's unlikely that the upstream will accept these changes due to potential cross-platform issues with NaN, but we don't encounter this problem in our case.

img_v3_02dt_1164255e-c0e3-40cd-870b-a94abb0f65cg

**Comment by sundy-li** [comment]
> It's unlikely that the upstream will accept these changes due to potential cross-platform issues with NaN, but we don't encounter this problem in our case.


You can make it a feature gate that defaults to be disabled.  Send a pr to upstream, let the maintainer decide whether to accept it or not.

**Comment by dqhl76** [comment]
> > It's unlikely that the upstream will accept these changes due to potential cross-platform issues with NaN, but we don't encounter this problem in our case.
> 
> You can make it a feature gate that defaults to be disabled. Send a pr to upstream, let the maintainer decide whether to accept it.

Ok, I will send a PR to upstream later

**Comment by dqhl76** [comment]
> You can make it a feature gate that defaults to be disabled. Send a pr to upstream, let the maintainer decide whether to accept it.

I opened a PR yesterday. Maintainers have some concerns about that. https://github.com/near/borsh-rs/pull/308

**Comment by sundy-li** [comment]
LGTM, maybe we can do better.
```
#[derive(Default, Clone, Copy)]
#[repr(transparent)]
struct F32(OrderedFloat<f32>);
struct F64(OrderedFloat<f64>);
```

then let's implement serde， ord， eq  for the new struct

And `AsRef` ， `from(f32)`, ...    for it.

**Comment by dqhl76** [comment]
Hi @sundy-li, Could you please take a look again?

**Comment by andylokandy** [comment]
I'm concerned about moving the whole lib into databend. As @sundy-li said, we can impl the custom serde logic for `F32` and `F64`.

Even if we decide to vendor the crate, it's better to give it its own crate rather than moving into `databend-common-base`.

**Comment by dqhl76** [comment]
Thanks for your review.

> I'm concerned about moving the whole lib into databend. As @sundy-li said, we can impl the custom serde logic for `F32` and `F64`.

If we implement these traits (like serde, Ord, Eq, AsRef, From(f32)) ourselves, we’re essentially just duplicating what the ordered-float library already does. In that case, we migh vendor the lib directly.

> Even if we decide to vendor the crate, it's better to give it its own crate rather than moving into `databend-common-base`.

It sounds good to me. I didn't put it into a crate becuase I thought it is a `one file crate`.

**Review by sundy-li — APPROVED** [review]

**Pull request merged**
